PennMobile/Course Alerts/Networking/CourseAlertNetworkManager.swift

```diff
@@ -0,0 +1,316 @@
+//
+//  CourseAlertNetworkManager.swift
+//  PennMobile
+//
+//  Created by Raunaq Singh on 10/25/20.
+//  Copyright © 2020 PennLabs. All rights reserved.
+//
+
+import Foundation
+import SwiftyJSON
+
+struct Response: Decodable {
+    let message: String
+    //let id: Int
+}
+
+
+class CourseAlertNetworkManager: NSObject, Requestable {
+    
+    static let instance = CourseAlertNetworkManager()
+    
+    let settingsURL = "https://penncoursealert.com/api/alert/settings/"
+    let coursesURL = "https://penncoursealert.com/api/alert/courses"
+    let registrationsURL = "https://penncoursealert.com/api/alert/registrations/"
+    
+    func getSearchedCourses(searchText:String, _ callback: @escaping (_ results: [CourseSection]?) -> ()) {
+        let urlStr = "\(coursesURL)/?search=\(searchText)"
+        let url = URL(string: urlStr.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!)!
+        let task = URLSession.shared.dataTask(with: url) { (data, response, error) in
+            if let data = data {
+                if let sectionsData = try? JSON(data).rawData() {
+                    let results = try? JSONDecoder().decode([CourseSection].self, from: sectionsData)
+                    callback(results)
+                    return
+                }
+            }
+            callback(nil)
+        }
+        task.resume()
+    }
+    
+    func getRegistrations(callback: @escaping ([CourseAlert]?) -> ()) {
+        makeGetRequestWithAccessToken(url: registrationsURL) { (data, response, error) in
+            guard let data = data, error == nil  else {
+                callback(nil)
+                return
+            }
+            
+            guard let registrations = try? JSONDecoder().decode([CourseAlert].self, from: data) else {
+                callback(nil)
+                return
+            }
+
+            callback(registrations)
+        }
+    }
+    
+    func getSettings(callback: @escaping (CourseAlertSettings?) -> ()) {
+        makeGetRequestWithAccessToken(url: settingsURL) { (data, response, error) in
+            guard let data = data, error == nil  else {
+                callback(nil)
+                return
+            }
+            
+            //print(String(decoding: data, as: UTF8.self))
+            
+            guard let settings = try? JSONDecoder().decode(CourseAlertSettings.self, from: data) else {
+                callback(nil)
+                return
+            }
```

Comment:
I store the settings locally so that every time they open the PCA page it doesn't have to wait to retrieve the settings until the page loads (sometimes it takes a second or two), instead it just uses the past setting first, then after new settings are retrieved, it updates the page if the setting was changed. I think that generally people will not be changing that setting too often, so really after the first time they enable PCA on mobile, it will pretty much stay that way, so this makes it a little smoother.

Reply:
awesome